### PR TITLE
Add diagnostic to flag unsupported usages of initializer APIs

### LINKED ISSUE #54381: Calling field initializers (input, output, queries, etc) should be errors when not used to declare a field in a directive
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

Yes

### Description

Assign viewChildren() to a variable instead of a property.
```
export class AppComponent {
  constructor() {
    // assign viewChildren() to a variable instead of a property
    const titles = viewChildren<string, ElementRef<HTMLHeadingElement>>(
      'title',
      {
        read: ElementRef,
      }
    );

    effect(() => {
      console.log(titles());
    });
  }
}
```

run ngc
`yarn run ngc -p tsconfig.json`

output
```
export class AppComponent {
  constructor() {
    // assign viewChildren() to a variable not a property
    const titles = viewChildren("title", {
      read: ElementRef,
    });
    effect(() => {
      console.log(titles());
    });
  }
  static {
    this.ɵfac = function AppComponent_Factory(t) {
      return new (t || AppComponent)();
    };
  }
  static {
    this.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({
      type: AppComponent,
      selectors: [["app-root"]],
      standalone: true,
      features: [i0.ɵɵStandaloneFeature],
      decls: 2,
      vars: 0,
      consts: [["title", ""]],
      template: function AppComponent_Template(rf, ctx) {
        if (rf & 1) {
          i0.ɵɵelement(0, "h1", null, 0);
        }
      },
    });
  }
}
```
the compile no generate viewQuery: function AppComponent_Query(rf, ctx) code.
I try another way also same
```
export class AppComponent {
  titles: Signal<readonly ElementRef<HTMLHeadingElement>[]>;

  constructor() {
    this.titles = viewChildren<string, ElementRef<HTMLHeadingElement>>(
      'title', { read: ElementRef, }
    );

    effect(() => {
      console.log(this.titles());
    });
  }
}
```
Not really sure if it's a bug or by design. Does the code style have to follow the required format for compilation?

### Please provide a link to a minimal reproduction of the bug

https://github.com/keatkeat87/Angular-viewChildren-not-be-compile-when-assign-to-variable

### Please provide the exception or error you saw

```true
no exception
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.2.0-rc.0
Node: 18.17.0
Package Manager: yarn 1.22.19
OS: win32 x64

Angular: 17.2.0-rc.1
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1702.0-rc.0
@angular-devkit/build-angular   17.2.0-rc.0
@angular-devkit/core            17.2.0-rc.0
@angular-devkit/schematics      17.2.0-rc.0
@angular/cli                    17.2.0-rc.0
@schematics/angular             17.2.0-rc.0
rxjs                            7.8.1
typescript                      5.3.3
zone.js                         0.14.3
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by crisbeto**
Adds a diagnostic that will flag any usages of initializer APIs outside of initializers. Includes the following changes:

### refactor(compiler-cli): add API to check if imports exist
Adds a couple of APIs to the `ImportedSymbolsTracker` that allow us to quickly check if a specific symbol is imported in a file.

### refactor(compiler-cli): introduce infrastructure for running additional checks against TypeScript files
Adds the new `SourceFileValidator` that will be used to check for file-level issues that may prevent Angular from working, like invoking the `input()` function outside of an initializer. Currently only one check is planned, but this setup will allow us to easily add more in the future.

### fix(compiler-cli): add diagnostic if initializer API is used outside of an initializer
Adds a rule that will produce a diagnostic when an initializer-based API is used outside of an initializer.

### fix(compiler-cli): report cases where initializer APIs are used in a non-directive class
Expands the check for initializer APIs to also flag when the function is called on a class that isn't a component or directive.

### refactor(compiler-cli): allow source file validator rule to produce a single diagnostic
Allows for `SourceFileValidatorRule.checkNode` to produce a single diagnostic. The most common case should be one diagnostic per node so this allows us to save some array allocations.

Fixes #54381.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  {functionName: 'contentChildren', owningModule: '@angular/core'},
+];
+
+/**
+ * Rule that flags any initializer APIs that are used outside of an initializer.
+ */
+export class InitializerApiUsageRule implements SourceFileValidatorRule {
+  constructor(
+      private reflector: ReflectionHost, private importedSymbolsTracker: ImportedSymbolsTracker) {}
+
```
Comment (crisbeto) [inline packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:36]: It's worth noting that this check currently doesn't capture cases like `input.call(null)`. I wanted to keep it simple for now and see if they even come up.

**Review by devversion — APPROVED** [review]
Nice 🎉

File: packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+const APIS_TO_CHECK: InitializerApiFunction[] = [
+  {functionName: 'input', owningModule: '@angular/core'},
+  {functionName: 'output', owningModule: '@angular/core'},
+  {functionName: 'outputFromObservable', owningModule: '@angular/core/rxjs-interop'},
+  {functionName: 'model', owningModule: '@angular/core'},
+  {functionName: 'viewChild', owningModule: '@angular/core'},
+  {functionName: 'viewChildren', owningModule: '@angular/core'},
+  {functionName: 'contentChild', owningModule: '@angular/core'},
+  {functionName: 'contentChildren', owningModule: '@angular/core'},
+];
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:28]: Yeah it's doable, although I guess we'll still have to remember to update this file if we add ned APIs.

File: packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    while (node.parent &&
+           (ts.isParenthesizedExpression(node.parent) || ts.isAsExpression(node.parent))) {
+      node = node.parent;
+    }
+
+    // Initializer functions are allowed to be used in the initializer.
+    if (!node.parent || !ts.isCallExpression(node) ||
+        (ts.isPropertyDeclaration(node.parent) && node.parent.initializer === node)) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:61]: Hmm, good question. I was under the impression that the handlers would flag cases like that, but I haven't checked it.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Review by atscott — APPROVED** [review]
Reviewed-for: public-api

File: packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:28]: I think we should do this, especially with the changes from the other PR. Can follow-up though

File: packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/validation/src/rules/initializer_api_usage_rule.ts:61]: Should we add it here?

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Comment by crisbeto** [comment]
This change should be ready to go now. I've rebased, addressed the feedback and expanded the check to also cover classes that aren't components/directives.

**Review by devversion — APPROVED** [review]

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 336916ceb07947bc78d4483f95d086e7f4841a30.

**Pull request closed without merging**